File: Leetcode/Typescript/2364. Count Number of Bad Pairs/solution.ts
// https://leetcode.com/problems/count-number-of-bad-pairs/description

function countBadPairs(nums: number[]): number {
  const pairNumberCount = new Map<number, number>();
  let badPairs = 0;

  for (let idx = 0; idx < nums.length; idx++) {
    // (j - i != nums[j] - nums[i]) can be (nums[i] - i != nums[j] - j)
    const pairNumber = nums[idx] - idx;
    badPairs += idx - (pairNumberCount.get(pairNumber) || 0);

    pairNumberCount.set(pairNumber, (pairNumberCount.get(pairNumber) || 0) + 1);
  }

  return badPairs;
}

File: Leetcode/Python/2364. Count Number of Bad Pairs/solution.py
# https://leetcode.com/problems/count-number-of-bad-pairs/description

class Solution:
    def countBadPairs(self, nums: List[int]) -> int:
        pair_count = defaultdict(int)
        bad_pairs = 0

        for idx in range(len(nums)):
            # (j - i != nums[j] - nums[i]) can be (nums[i] - i != nums[j] - j)
            pair_number = nums[idx] - idx
            bad_pairs += idx - pair_count[pair_number]

            pair_count[pair_number] += 1

        return bad_pairs
